### PR TITLE
[OGR provider] Handles layer of geometry type unknown with null geometries

### LINKED ISSUE #50215: GML import is missing one feature without geometry
### What is the bug or the crash?

GML import is missing one feature without geometry.

There are 1759 features in total, but one feature without geometry is not shown for the import dialog:

image

### Steps to reproduce the issue

Drag'n'drop the following GML into QGIS: [xplansynwfs-4.zip](https://github.com/qgis/QGIS/files/9574377/xplansynwfs-4.zip)

The following feature is missing (the first one in the document):

  ```xml
<wfs:member>
    <xplan:FP_BebauungsFlaeche xmlns:xplan="http://www.deegree.org/xplanung/1/0" gml:id="XPLAN_FP_BEBAUUNGSFLAECHE_cc4f50a8-a2e7-4c5c-bc5d-c5b82d5079b0">
      <xplan:gmlId>XPLAN_FP_BEBAUUNGSFLAECHE_cc4f50a8-a2e7-4c5c-bc5d-c5b82d5079b0</xplan:gmlId>
      <xplan:xpVersion>5.1</xplan:xpVersion>
      <xplan:xpPlanType>FP_Plan</xplan:xpPlanType>
      <xplan:xpPlanName>FP Krefeld (Urschrift)</xplan:xpPlanName>
      <xplan:xplanMgrPlanId>144</xplan:xplanMgrPlanId>
      <xplan:ebene>0</xplan:ebene>
      <xplan:gehoertZuBereich>[XPLAN_FP_BEREICH_8f95f31c-fb74-40d6-a77d-78666cbff16e]</xplan:gehoertZuBereich>
      <xplan:rechtscharakter>Darstellung</xplan:rechtscharakter>
      <xplan:rechtscharakterCode>1000</xplan:rechtscharakterCode>
      <xplan:flaechenschluss>true</xplan:flaechenschluss>
      <xplan:allgArtDerBaulNutzung>GemischteBauflaeche</xplan:allgArtDerBaulNutzung>
      <xplan:allgArtDerBaulNutzungCode>2000</xplan:allgArtDerBaulNutzungCode>
      <xplan:besondereArtDerBaulNutzung>Kerngebiet</xplan:besondereArtDerBaulNutzung>
      <xplan:besondereArtDerBaulNutzungCode>1600</xplan:besondereArtDerBaulNutzungCode>
    </xplan:FP_BebauungsFlaeche>
  </wfs:member>
```

### Versions

QGIS version | 3.27.0-Master | QGIS code revision | 9c4d0a2064
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.10.4
GDAL/OGR version | 3.4.1
PROJ version | 8.2.1
EPSG Registry database version | v10.041 (2021-12-03)
GEOS version | 3.10.2-CAPI-1.16.0
SQLite version | 3.37.2
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.4
QScintilla2 version | 2.11.6
OS version | Ubuntu 22.04.1 LTS
  |   |   |  
Active Python plugins
sagaprovider | 2.12.99
grassprovider | 2.12.99
MetaSearch | 0.3.6
processing | 2.12.99
db_manager | 0.1.20

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rouault**
Fixes #50215

Layers with a wkbUnknown OGR layer geometry types are first scanned to figure out which actualy geometry types are present, to expose sublayers filtered with the |geometryType=xxxx URI parameter.

That hint is then used by getFeatureCount() to only count features with such geometry type, and also by the feature iterator. A consequence of that is that features with null geometries are excluded.

When there is a single geometry type, we could then just use the general purpose code paths, ie not actually filter on geometry type, so do that by adding a |uniqueGeometryType=yes URI parameter. Note that this may not be fully bullet-proof if adding such layer to a project, and later the datasource is modified to add geometries of another type.

And note also that there is another issue, pre-existing to this PR and not addressed by it, if a layer contains a feature with a geometry of type A, one of type B and a feature with no geometry. QGIS will never report the feature with no geometry.

Related GDAL fix specifically for the original use case of #50215 in https://github.com/OSGeo/gdal/pull/6377

### CODE REVIEW

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -401,6 +401,13 @@ QgsOgrProvider::QgsOgrProvider( QString const &uri, const ProviderOptions &optio
               mOgrGeometryTypeFilter,
               mOpenOptions );
 
+
+  QVariantMap parts = QgsOgrProviderMetadata().decodeUri( uri );
```
Comment (elpaso) [inline src/core/providers/ogr/qgsogrprovider.cpp:405]:
```suggestion
  const QVariantMap parts = QgsOgrProviderMetadata().decodeUri( uri );
```

**Review by elpaso — APPROVED** [review]

File: src/core/providers/ogr/qgsogrprovider.cpp (hunk 1 above)
Comment (rouault) [reply src/core/providers/ogr/qgsogrprovider.cpp]: done

**Comment by pathmapper** [comment]
Thanks @rouault @elpaso 

Confirmed that with this fix all 1759 features of the sample case in https://github.com/qgis/QGIS/issues/50215 are imported.

The import dialog still shows 1758 features but if I got it right, with https://github.com/OSGeo/gdal/pull/6377 the geometry type of the sample case will be detected as `wkbPolygon` so the data is then directly imported without showing the import dialog.

**Comment by rouault** [comment]
> The import dialog still shows 1758 features

addressed by 6a7f176ecacbee86afd6677a899ccdddc7356f2c

**Comment by pathmapper** [comment]
> addressed by https://github.com/qgis/QGIS/commit/6a7f176ecacbee86afd6677a899ccdddc7356f2c

Works great :tada: 

image

**Comment by pathmapper** [comment]
For a GML with only one layer containing only features with no geometry the import dialog feature count = 0 (the import of the features works).

Test data: [xplansynwfs.zip](https://github.com/qgis/QGIS/files/9592338/xplansynwfs.zip)

image

@rouault is this the same issue you've mentioned in the PR description?

> if a layer contains a feature with a geometry of type A, one of type B and a feature with no geometry. QGIS will never report the feature with no geometry.

**Comment by rouault** [comment]
> For a GML with only one layer containing only features with no geometry the import dialog feature count = 0 (the import of the features works).

addressed per e445f2b49baa961aaab8c4f6dc02aa8e86d1f8bb

**Comment by pathmapper** [comment]
> addressed per https://github.com/qgis/QGIS/commit/e445f2b49baa961aaab8c4f6dc02aa8e86d1f8bb

Nice - tested and confirmed!

**Review by lbartoletti — APPROVED** [review]

**Pull request merged**